### ISSUE TITLE
findUnique return type does not include null

### ISSUE THREAD

**Issue by scefali** [prisma/prisma#10903]
## Problem

The docs say `findUnique` can return zero or one items that match a query. The issue is that the Typescript return type of `findUnique` is simply a promise of an instance of the model you are querying and does not account for when there isn't a match and the result is `null`. This can lead to bugs like `Cannot read properties of null (reading 'password')` unless you check the result of `findUnique` isn't null.

See below where `user` is typed `User` instead of `User | null`.

![Screen Shot 2021-12-29 at 3 44 27 PM](https://user-images.githubusercontent.com/8533851/147711047-860dc858-6c76-416a-8df7-3cddde355ab0.png)

## Suggested solution

A setting or something so that I can type the return type of `findUnique` to also include `null` to force proper type checking. Another alternative is to throw an error if the query doesn't find a match which is how Django does it. But maybe such a setting exists and it's not documented?

## Alternatives

I know I could just remember to check the output of `findUnique` but the whole point of Typescript is so I don't have to remember things like this.

**Comment by sabinadams** [comment]
`Another alternative is to throw an error if the query doesn't find a match which is how Django does it. But maybe such a setting exists and it's not documented?`

There is a setting for this called `rejectOnNotFound` that will throw an error if no matches are found.

**Comment by sabinadams** [comment]
> ## Problem
> The docs say `findUnique` can return zero or one items that match a query. The issue is that the Typescript return type of `findUnique` is simply a promise of an instance of the model you are querying and does not account for when there isn't a match and the result is `null`. This can lead to bugs like `Cannot read properties of null (reading 'password')` unless you check the result of `findUnique` isn't null.
> 
> See below where `user` is typed `User` instead of `User | null`.
> 
> ![Screen Shot 2021-12-29 at 3 44 27 PM](https://user-images.githubusercontent.com/8533851/147711047-860dc858-6c76-416a-8df7-3cddde355ab0.png)
> 
> ## Suggested solution
> A setting or something so that I can type the return type of `findUnique` to also include `null` to force proper type checking. Another alternative is to throw an error if the query doesn't find a match which is how Django does it. But maybe such a setting exists and it's not documented?
> 
> ## Alternatives
> I know I could just remember to check the output of `findUnique` but the whole point of Typescript is so I don't have to remember things like this.

There is a setting for this called rejectOnNotFound that will throw an error if no matches are found.
https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#rejectonnotfound

**Comment by scefali** [comment]
@sabinadams Great, this is exactly what I'm looking for. Thanks!

**Comment by omidh28** [comment]
Why not overload `findUnique` so it would include `null` if `rejectOnNotFound` is explicitly set?

**Comment by MeoMix** [comment]
> I also strongly believe this issue should not be closed. My understanding is that prisma's main feature is it's advanced type safety. Here we have a function that returns `null | Type` but does not label itself as such. `findUnique` is simply not typed correctly in Prisma and `rejectOnNotFound` is a hatch - not a fix.

+1

I want a solution where `rejectOnNotFound` is false and the return type of `findUnique<T>` is `T | null`

**Comment by rosskuyper** [comment]
For those who find this, this can be achieved with the correct tsconfig settings, more info [here](https://github.com/prisma/prisma/discussions/10075#discussioncomment-1581284)

tl;dr - requires strict mode or strict null checks enabled
